egs/librispeech/asr/simple_v1/mmi_bigram_embeddings_decode.py

```diff
@@ -0,0 +1,402 @@
+
+
+def decode(dataloader: torch.utils.data.DataLoader, model: AcousticModel,
+           second_pass_model: AcousticModel,
+           ctc_topo: k2.Fsa,
+           max_phone_id: int,
+           device: Union[str, torch.device], LG: Fsa, symbols: SymbolTable,
+           enable_second_pass_decoding: bool
+           ):
+    tot_num_cuts = len(dataloader.dataset.cuts)
+    num_cuts = 0
+    results = []  # a list of pair (ref_words, hyp_words)
+    for batch_idx, batch in enumerate(dataloader):
+        feature = batch['features']
+        supervisions = batch['supervisions']
+        supervision_segments = torch.stack(
+            (supervisions['sequence_idx'],
+             torch.floor_divide(supervisions['start_frame'],
+                                model.subsampling_factor),
+             torch.floor_divide(supervisions['num_frames'],
+                                model.subsampling_factor)), 1).to(torch.int32)
+        indices = torch.argsort(supervision_segments[:, 2], descending=True)
+        supervision_segments = supervision_segments[indices]
+        texts = supervisions['text']
+        assert feature.ndim == 3
+
+        feature = feature.to(device)
+        # at entry, feature is [N, T, C]
+        feature = feature.permute(0, 2, 1)  # now feature is [N, C, T]
+        with torch.no_grad():
+            nnet_output, nnet_output_2nd = model(feature)
+        # nnet_output is [N, C, T]
+        nnet_output = nnet_output.permute(0, 2, 1)  # now nnet_output is [N, T, C]
+        nnet_output_2nd = nnet_output_2nd.permute(0, 2, 1)  # now nnet_output_2nd is [N, T, C]
+
+        #  blank_bias = -3.0
+        #  nnet_output[:, :, 0] += blank_bias
+
+        dense_fsa_vec = k2.DenseFsaVec(nnet_output, supervision_segments)
+        dense_fsa_vec_2nd = k2.DenseFsaVec(nnet_output_2nd, supervision_segments)
+        # assert LG.is_cuda()
+        assert LG.device == nnet_output.device, \
+            f"Check failed: LG.device ({LG.device}) == nnet_output.device ({nnet_output.device})"
+        # TODO(haowen): with a small `beam`, we may get empty `target_graph`,
+        # thus `tot_scores` will be `inf`. Definitely we need to handle this later.
+        lattices = k2.intersect_dense_pruned(LG, dense_fsa_vec, 20.0, 7.0, 30,
+                                             10000)
+
+        if not enable_second_pass_decoding:
+            # lattices = k2.intersect_dense(LG, dense_fsa_vec, 10.0)
+            best_paths = k2.shortest_path(lattices, use_double_scores=True)
+        else:
+            paths = get_paths(lats=lattices, num_paths=10)
+            word_fsas, seq_to_path_shape = get_word_fsas(lattices, paths)
+            replicated_lats = k2.index(lattices, seq_to_path_shape.row_ids(1))
+            word_lats = k2.compose(replicated_lats,
+                                   word_fsas,
+                                   treat_epsilons_specially=False)
+            tot_scores_1st = word_lats.get_tot_scores(use_double_scores=True,
```

Comment:
From https://github.com/k2-fsa/snowfall/pull/106#discussion_r595941948

It is this line that **somtimes** causes the check failure from https://github.com/k2-fsa/k2/blob/171ddc17509c0de5e9f7dcc4efeed9c712830233/k2/csrc/fsa_utils.cu#L699
```cpp
K2_CHECK_GT(dest_state, state_idx01);
```

Here is the call stack from the Python side
```
Traceback (most recent call last):
  File "./mmi_bigram_embeddings_decode.py", line 404, in <module>
    main()
  File "./mmi_bigram_embeddings_decode.py", line 370, in main
    results = decode(dataloader=test_dl,
  File "./mmi_bigram_embeddings_decode.py", line 97, in decode
    tot_scores_1st = word_lats.get_tot_scores(use_double_scores=True,
  File "/root/fangjun/open-source/k2/k2/python/k2/fsa.py", line 598, in get_tot_scores
    tot_scores = k2.autograd._GetTotScoresFunction.apply(
  File "/root/fangjun/open-source/k2/k2/python/k2/autograd.py", line 49, in forward
    tot_scores = fsas._get_tot_scores(use_double_scores=use_double_scores,
  File "/root/fangjun/open-source/k2/k2/python/k2/fsa.py", line 577, in _get_tot_scores
    forward_scores = self._get_forward_scores(use_double_scores,
  File "/root/fangjun/open-source/k2/k2/python/k2/fsa.py", line 526, in _get_forward_scores
    state_batches=self._get_state_batches(),
  File "/root/fangjun/open-source/k2/k2/python/k2/fsa.py", line 433, in _get_state_batches
    cache[name] = _k2.get_state_batches(self.arcs, transpose=True)
```

Regarding
> That is essentially asserting that the FSA is top-sorted and acyclic. We'd have to consider where the FSA came from

The FsaVec `word_lats` is from `k2.compose(first_pass_lats, word_fsas)`,
where `word_fsas` is from
https://github.com/k2-fsa/snowfall/blob/8e47582d6dfaf818bdb3a63da08dbaa5ca915a2a/snowfall/decoding/rescore.py#L59-L62

After printing the properties of the problematic `word_lats`, it shows
```
"Valid|Nonempty|MaybeAccessible"
```

which is different from the normal `word_lats` with the following properties
```
"Valid|Nonempty|TopSorted|TopSortedAndAcyclic|MaybeAccessible"
```

---

I am looking into it.